core/src/main/java/org/aerogear/mobile/core/http/OkHttpResponse.java

```diff
@@ -66,8 +87,10 @@ public void waitForCompletionAndClose() {
             requestCompleteLatch.await(DEFAULT_TIMEOUT, TimeUnit.SECONDS);
             //If a completion Handler was set before this wait then we need to make sure that
             // gets called before we free these resources.
-            if (completionHandler != null) {
+            if (completionHandler != null && requestError == null) {
```

Comment:
Like I said before if we have a `completionHandler` it should be always called, not only on success

Reply:
Hmm, wouldn't that make the error handler pointless? Because then you would have to check for errors in your completion handler.

Reply:
@danielpassos I was thinking about the same, but if you read the javadoc of the `onComplete` method:

> /**
     * Creates a callback to be called when the response is finished successfully.
     *
     * The response is delivered immediately if the request has already been finished.
     */

It says that it is only called when the response is finished successfully.  

If this is still the case, then I think it shouldn't be called when there is an error.

I think the name of the method is a bit confusing. I agree that if it is called `onComplete`, then it should be called all the time. Otherwise it's better rename it to `onSuccess`.

Reply:
@wei-lee i'll update this PR soon to include `onSuccess`, `onError` and `onComplete`.

Reply:
@pb82 👍